va/validation-authority.go

```diff
@@ -498,20 +498,28 @@ func (va *ValidationAuthorityImpl) validateDNS01(ctx context.Context, identifier
 }
 
 func (va *ValidationAuthorityImpl) checkCAA(ctx context.Context, identifier core.AcmeIdentifier) *probs.ProblemDetails {
-	// Check CAA records for the requested identifier
-	present, valid, err := va.checkCAARecords(ctx, identifier)
-	if err != nil {
-		return bdns.ProblemDetailsFromDNSError(err)
-	}
-	// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-	va.log.AuditInfo(fmt.Sprintf("Checked CAA records for %s, [Present: %t, Valid for issuance: %t]", identifier.Value, present, valid))
-	if !valid {
-		return &probs.ProblemDetails{
-			Type:   probs.ConnectionProblem,
-			Detail: fmt.Sprintf("CAA record for %s prevents issuance", identifier.Value),
+	var prob *probs.ProblemDetails
+	if va.caaClient != nil {
+		prob = va.checkCAAService(ctx, identifier)
+	} else {
+		present, valid, err := va.checkCAARecords(ctx, identifier)
+		if err != nil {
+			prob = bdns.ProblemDetailsFromDNSError(err)
+		} else {
+			// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+			va.log.AuditInfo(fmt.Sprintf("Checked CAA records for %s, [Present: %t, Valid for issuance: %t]", identifier.Value, present, valid))
+			if !valid {
+				prob = &probs.ProblemDetails{
+					Type:   probs.CAAProblem,
+					Detail: fmt.Sprintf("CAA record for %s prevents issuance", identifier.Value),
+				}
+			}
 		}
 	}
-	return nil
+	if va.caaDR != nil && prob != nil && prob.Type == probs.ConnectionProblem {
```

Comment:
I think this is required to keep the compatibility between the wrappers for the built-in and gRPC CAA checking services. Once the built-in service is removed this can be simplified to not use `ProblemDetails` in this way but I think doing it now would add a bunch more churn to this changeset that would still need to be reverted to some degree once we remove the built-in.